### PR TITLE
Various Updates, including some for functionality around knative v0.4.0

### PR DESCRIPTION
**Pull request by lberk**
Hi, please see the commits to help get things working on knative v0.4.0

This includes fixing the autoscaling section by adjusting our first definition of the autoscale-configmap so that the rest of the section and proceed unbroken (v0.4.0 moves the previously existing variables into a comment).

We also have moved from the (deprecated) 'pinned.revisionName' versioning to 'release.revisions' moving forward.

One aspect this does not fix, is how deployment names are now unordered and randomly generated.

### CODE REVIEW

File: 01-basics/knative/service-pinned-rev1.yaml [hunk 1]
```diff
@@ -3,8 +3,8 @@ kind: Service
 metadata:
   name: greeter
```
Comment (bbrowning) [inline 01-basics/knative/service-pinned-rev1.yaml:4]: missing namespace

File: 03-scaling/knative/service-10.yaml [hunk 2]
```diff
@@ -2,6 +2,7 @@ apiVersion: serving.knative.dev/v1alpha1
 kind: Service
 metadata:
   name: greeter
+  namespace: knativetutorial
   annotations:
     # Target 10 in-flight-requests per pod.
     autoscaling.knative.dev/target: "10"
```
Comment (bbrowning) [inline 03-scaling/knative/service-10.yaml:8]: This annotation is in the wrong place but #74 fixes that. Just be aware a merge conflict may need to be resolved in this file given how close these two changes are together.

File: documentation/modules/ROOT/pages/setup.adoc [hunk 3]
```diff
@@ -81,8 +81,8 @@ The `work` folder in `$TUTORIAL_HOME` can be used to download the demo applicati
 
 This tutorial was developed and tested with:
 
-- Knative `v0.3.0`
-- minikube `v0.34.1`
+- Knative `v0.4.0`
```
Comment (bbrowning) [inline documentation/modules/ROOT/pages/setup.adoc:84]: You also need to update knative-version in documentation/modules/ROOT/pages/_attributes.adoc .

But, there's a wrinkle - we shouldn't just assume build/serving/eventing are all the exact same version. Serving is on 0.4.1, for example - patch releases get cut for each automatically based on need. So, we should separate out the versions of build/serving/eventing and be explicit with each. That could wait for a subsequent PR, but noting it here for now.

**Review by bbrowning — COMMENTED** [review]
I didn't do a complete review of all the changes, but did notice a couple of things. Thanks for tackling this!

File: 01-basics/knative/service-pinned-rev1.yaml (hunk 1 above)
Comment (lberk) [reply 01-basics/knative/service-pinned-rev1.yaml:4]: ah nice catch, thanks! Fixed.

File: 03-scaling/knative/service-10.yaml (hunk 2 above)
Comment (lberk) [reply 03-scaling/knative/service-10.yaml:8]: Thanks for the heads up

File: 05-eventing/knative/event-source-svc.yaml [hunk 4]
```diff
@@ -2,8 +2,9 @@ apiVersion: sources.eventing.knative.dev/v1alpha1
 kind: CronJobSource
 metadata:
   name: event-greeter-cronjob-source
+  namespace: knativetutorial
 spec:
-  schedule: "* * * * *"
+  schedule: "*/1 * * * *"
```
Comment (syedriko) [inline 05-eventing/knative/event-source-svc.yaml:7]: "* * * * *" means "every minute", I don't think there's anything to fix here.

File: 05-eventing/knative/event-source.yaml [hunk 5]
```diff
@@ -2,8 +2,9 @@ apiVersion: sources.eventing.knative.dev/v1alpha1
 kind: CronJobSource
 metadata:
   name: event-greeter-cronjob-source
+  namespace: knativetutorial
 spec:
-  schedule: "* * * * *"
+  schedule: "*/1 * * * *"
```
Comment (syedriko) [inline 05-eventing/knative/event-source.yaml:7]: "* * * * *" means "every minute", I don't think there's anything to fix here.

File: documentation/modules/ROOT/pages/05eventing/eventing-src-sub.adoc [hunk 6]
```diff
@@ -84,8 +85,9 @@ apiVersion: sources.eventing.knative.dev/v1alpha1
 kind: CronJobSource #<1>
 metadata:
   name: event-greeter-cronjob-source
+  namespace: knativetutorial
 spec:
-  schedule: "* * * * *"
+  schedule: "*/1 * * * *"
```
Comment (syedriko) [inline documentation/modules/ROOT/pages/05eventing/eventing-src-sub.adoc:90]: "* * * * *" means "every minute", I don't think there's anything to fix here.

File: documentation/modules/ROOT/pages/05eventing/eventing-src-svc.adoc [hunk 7]
```diff
@@ -23,10 +117,11 @@ apiVersion: sources.eventing.knative.dev/v1alpha1
 kind: CronJobSource #<1>
 metadata:
   name: event-greeter-cronjob-source
+  namespace: knativetutorial
 spec:
-  schedule: "* * * * *"
+  schedule: "*/1 * * * *"
```
Comment (syedriko) [inline documentation/modules/ROOT/pages/05eventing/eventing-src-svc.adoc:122]: "* * * * *" means "every minute", I don't think there's anything to fix here.

File: documentation/modules/ROOT/pages/setup.adoc (hunk 3 above)
Comment (kameshsampath) [reply documentation/modules/ROOT/pages/setup.adoc:84]: @bbrowning  thanks for the head up, in earlier versions all were same, since its getting changed will have it updated.

File: 03-scaling/knative/service-10.yaml (hunk 2 above)
Comment (kameshsampath) [reply 03-scaling/knative/service-10.yaml:8]: @lberk  @bbrowning - I usually avoid the need to add `namespace` to the resource files 

- all the scripts and commands in the tutorial has `--namespace` or `-n` to the kubectl command that ensures what NS they are deployed to 
- Also giving the users the flexibility to change the namespace if they wish to

Comment (kameshsampath) [reply 03-scaling/knative/service-10.yaml:8]: > This annotation is in the wrong place but #74 fixes that. Just be aware a merge conflict may need to be resolved in this file given how close these two changes are together.

this is already merged and  @lberk has rebased it

Comment (lberk) [reply 03-scaling/knative/service-10.yaml:8]: @kameshsampath I can definitely appreciate that approach with allowing users to specify the namespace themselves, and I can also appreciate that this kind of change might land in the realm of bikeshedding.  However, when running through the tutorial I made multiple typo's (maybe that's just my own fault) leading to having to check where I'd applied the actual resource when things didn't work as advertised.  Even tab-completing didn't quite work that well because the namespace `knativetutorial` also competes with knative's own namespaces (`knative-serving`, `knative-building`, `knative-eventing`, `knative-sources`, etc).  If our goal for the tutorial is to make it as easy as possible for new users to get up to speed, I'd recommend specifying this in the yaml files, to make running through the tutorial as easy as possible.

Comment (syedriko) [reply 03-scaling/knative/service-10.yaml:8]: @lberk @kameshsampath I tend to agree with Lukas here, the less we make ppl type, the better. I for one copy-pasted the commands from the tutorial. It would have taken me twice the time if I had to type them.
What if we employ the default namespace feature of kubectl? Set it at the start, reset back to what it was in the end?
https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/#setting-the-namespace-preference:
```kubectl config set-context $(kubectl config current-context) --namespace=<insert-namespace-name-here>```

File: documentation/modules/ROOT/pages/03scaling.adoc [hunk 8]
```diff
@@ -18,6 +18,48 @@ include::partial$prereq-cli.adoc[]
...
+[source,bash,subs="+macros,+attributes"]
+----
+cd $TUTORIAL_HOME/03-serving/knative
+----
+
+The service can be deployed using the command:
+
+[source,bash,subs="+macros,+attributes"]
+----
+kubectl apply -f link:{github-repo}/{serving-repo}/knative/autoscaling-configmap.yaml[autoscaling-configmap.yaml]
```
Comment (rhuss) [inline documentation/modules/ROOT/pages/03scaling.adoc:53]: Should be `knative-serving-repo`, not `serving-repo`

File: documentation/modules/ROOT/pages/03scaling.adoc [hunk 9]
```diff
@@ -18,6 +18,48 @@ include::partial$prereq-cli.adoc[]
...
+----
+
+.(OR)
+
+[source,bash,subs="+macros,+attributes"]
+----
+oc apply -f link:{github-repo}/{serving-repo}/knative/autoscaling-configmap.yaml[autoscaling-configmap.yaml]
```
Comment (rhuss) [inline documentation/modules/ROOT/pages/03scaling.adoc:60]: `knative-serving-repo`

File: 03-scaling/knative/service-10.yaml (hunk 2 above)
Comment (kameshsampath) [reply 03-scaling/knative/service-10.yaml:8]: @lberk  @syedriko  - the `kubens` is your friend. I have quoted this tool in the pre-req and I use it heavily I am seeing it can also be installed via `krew`, so if you ppl are lazy typing namespaces then they an try setting it first ( I can add a tip to the doc) and then start typing other commands, IMHO using `--namespace` is always the k8s best practice for user clarity.

File: documentation/modules/ROOT/pages/03scaling.adoc (hunk 8 above)
Comment (kameshsampath) [reply documentation/modules/ROOT/pages/03scaling.adoc:53]: @rhuss  I am going to add few more variables based on @bbrowning  comment above for and that will fix the issue

File: documentation/modules/ROOT/pages/03scaling.adoc (hunk 9 above)
Comment (kameshsampath) [reply documentation/modules/ROOT/pages/03scaling.adoc:60]: same as above

File: 03-scaling/knative/service-10.yaml (hunk 2 above)
Comment (syedriko) [reply 03-scaling/knative/service-10.yaml:8]: @kameshsampath Thanks for the pointers to krew.dev and kubectx. However, I simply followed the tutorial and did not find a single mention of kubens' use and therefore did not use it. I even suggested to remove it from the prereqs in https://github.com/redhat-developer-demos/knative-tutorial/issues/80. Instead, I found many repetitive instructions for -n knativetutorial, which looked ripe for optimization.

Comment (kameshsampath) [reply 03-scaling/knative/service-10.yaml:8]: yeah thats my bad, as told will add `Tip` in the next release, though i dont want user to be forced to use that.

File: documentation/modules/ROOT/pages/setup.adoc [hunk 10]
```diff
@@ -105,8 +105,7 @@ minikube profile knative \#<1>
 minikube start -p knative --memory=8192 --cpus=4 \#<2> 
   --kubernetes-version=v1.12.0 \#<3> 
   --vm-driver=hyperkit \#<4>  
-  --disk-size=50g \
-  --extra-config=apiserver.enable-admission-plugins="LimitRanger,NamespaceExists,NamespaceLifecycle,ResourceQuota,ServiceAccount,DefaultStorageClass,MutatingAdmissionWebhook"
```
Comment (kameshsampath) [inline documentation/modules/ROOT/pages/setup.adoc:109]: why was this deleted? IIRC only `MutatingAdmissionWebhook` might become default with latest k8s.

**Comment by kameshsampath** [comment]
@lberk  - i need to skip some commits from this as they changing one v0.4.0 in attributes will update all, I already have fixes for them, will try to reuse them.

I have also re-ordered chapter numbers, so its worth you do a rebase before creating separate PR

Would you  mind splitting https://github.com/redhat-developer-demos/knative-tutorial/pull/77/commits/3b1bf62ffe2b316c6d2ea193900b115f336f2363 the commit as separate PR , I can then merge it as is now merging this whole PR will cause conflict.

File: documentation/modules/ROOT/pages/setup.adoc (hunk 10 above)
Comment (syedriko) [reply documentation/modules/ROOT/pages/setup.adoc:109]: Because only NamespaceExists is not on the list of enabled by default admission controllers, the rest of them are: https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#which-plugins-are-enabled-by-default. And things seem to work just fine even without NamespaceExists mentioned.

File: 03-scaling/knative/service-10.yaml (hunk 2 above)
Comment (lberk) [reply 03-scaling/knative/service-10.yaml:8]: @kameshsampath thanks as well for the notes about `kubens` but again, I'd reiterate that for the purposes of a tutorial, limiting the delta between their current setup and executing the contents as easily as possible is worth prioritising.  Are we teaching folks about knative? or krew, kubens, and knative and a variety of auxiliary tools?  If both @syedriko and I missed it, is it really that obvious to others?  Anyways, I'll rebase/fix the remaining sections as you suggest.

File: documentation/modules/ROOT/pages/setup.adoc (hunk 10 above)
Comment (lberk) [reply documentation/modules/ROOT/pages/setup.adoc:109]: @kameshsampath I didn't need it for my run through the tutorial, but if we'd like to keep them explicitly listed that's fine

**Comment by lberk** [comment]
> @lberk - i need to skip some commits from this as they changing one v0.4.0 in attributes will update all, I already have fixes for them, will try to reuse them.
> I have also re-ordered chapter numbers, so its worth you do a rebase before creating separate PR
> Would you mind splitting [3b1bf62](https://github.com/redhat-developer-demos/knative-tutorial/commit/3b1bf62ffe2b316c6d2ea193900b115f336f2363) the commit as separate PR , I can then merge it as is now merging this whole PR will cause conflict.

Ok, sure thing! What else were you looking to explicitly merge? From the comments above it seems like you a variety of things dropped?  Looking to make things as seamless & easy as possible for the contributions.

File: 03-scaling/knative/service-10.yaml (hunk 2 above)
Comment (kameshsampath) [reply 03-scaling/knative/service-10.yaml:8]: i have added the tip in the latest doc changes about `kubens`, thats the hard part about Kubernetes - lots of thing need to be learned in parallel :( sorry about that

**Comment by lberk** [comment]
New PR in https://github.com/redhat-developer-demos/knative-tutorial/pull/89

**Comment by kameshsampath** [comment]
addressed various fixes of other PRs, close this in favour of others.
